gax/src/main/java/com/google/api/gax/batching/NonBlockingSemaphore.java

```diff
@@ -31,37 +31,59 @@
 package com.google.api.gax.batching;
 
 import com.google.common.base.Preconditions;
-import java.util.concurrent.atomic.AtomicLong;
 
 /** A {@link Semaphore64} that immediately returns with failure if permits are not available. */
 class NonBlockingSemaphore implements Semaphore64 {
-  private final AtomicLong currentPermits;
+  private long currentPermits;
+  private long limit;
 
   private static void checkNotNegative(long l) {
     Preconditions.checkArgument(l >= 0, "negative permits not allowed: %s", l);
   }
 
   NonBlockingSemaphore(long permits) {
     checkNotNegative(permits);
-    this.currentPermits = new AtomicLong(permits);
+    this.currentPermits = permits;
+    this.limit = permits;
   }
 
-  public void release(long permits) {
+  public synchronized void release(long permits) {
```

Comment:
Using `synchronized` makes this method (and the others) blocking. Notice that the class is called `NonBlockingSemaphore`, so it must not block to server its purpose. We must use non-blocking synchronization algorithms here (which a usually very hard to implement; blocking synchronization is hard, non-blocking synchronization is even harder).